File: src/logger.ts
const LOG_LEVELS = { debug: 0, info: 1, warn: 2, error: 3 } as const;
type LogLevel = keyof typeof LOG_LEVELS;

let currentLevel: LogLevel = "info";

export function setLogLevel(level: LogLevel) {
  currentLevel = level;
}

function shouldLog(level: LogLevel): boolean {
  return LOG_LEVELS[level] >= LOG_LEVELS[currentLevel];
}

function formatTimestamp(): string {
  return new Date().toISOString();
}

function log(level: LogLevel, scope: string, message: string, data?: Record<string, unknown>) {
  if (!shouldLog(level)) return;

  const entry = {
    ts: formatTimestamp(),
    level,
    scope,
    msg: message,
    ...(data ? { data } : {}),
  };

  const line = JSON.stringify(entry);

  switch (level) {
    case "error":
      console.error(line);
      break;
    case "warn":
      console.warn(line);
      break;
    default:
      console.log(line);
  }
}

export function createLogger(scope: string) {
  return {
    debug: (msg: string, data?: Record<string, unknown>) => log("debug", scope, msg, data),
    info: (msg: string, data?: Record<string, unknown>) => log("info", scope, msg, data),
    warn: (msg: string, data?: Record<string, unknown>) => log("warn", scope, msg, data),
    error: (msg: string, data?: Record<string, unknown>) => log("error", scope, msg, data),
  };
}

File: src/utils/logger.py
"""Structured JSON logger matching the TS implementation's output format."""

from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone

LOG_LEVELS = {"debug": 0, "info": 1, "warn": 2, "error": 3}

_current_level: str = os.environ.get("LOG_LEVEL", "info").lower()


def set_log_level(level: str) -> None:
    global _current_level
    _current_level = level


def _should_log(level: str) -> bool:
    return LOG_LEVELS.get(level, 1) >= LOG_LEVELS.get(_current_level, 1)


def _log(level: str, scope: str, msg: str, data: dict | None = None) -> None:
    if not _should_log(level):
        return

    entry: dict = {
        "ts": datetime.now(timezone.utc).isoformat(),
        "level": level,
        "scope": scope,
        "msg": msg,
    }
    if data:
        entry["data"] = data

    line = json.dumps(entry, ensure_ascii=False)

    print(line, file=sys.stderr)


class Logger:
    __slots__ = ("_scope",)

    def __init__(self, scope: str) -> None:
        self._scope = scope

    def debug(self, msg: str, data: dict | None = None) -> None:
        _log("debug", self._scope, msg, data)

    def info(self, msg: str, data: dict | None = None) -> None:
        _log("info", self._scope, msg, data)

    def warn(self, msg: str, data: dict | None = None) -> None:
        _log("warn", self._scope, msg, data)

    def error(self, msg: str, data: dict | None = None) -> None:
        _log("error", self._scope, msg, data)


def create_logger(scope: str) -> Logger:
    return Logger(scope)
